plugins/overwatch.rb

```diff
@@ -0,0 +1,660 @@
+# Overwatch plugin - Active Routing and Session Management
+#
+# Author: sn0wfa11 - jhale85446[at]gmail.com
+#
+# Session stepping and logging functionality borrowed on HDM's Beholder plugin.
+
+require 'fileutils'
+
+module Msf
+
+class Plugin::Overwatch < Msf::Plugin
+
+  # Worker Thread
+  #
+  # @return [void] A useful return value is not expected here
+
+  class OverwatchWorker
+    attr_accessor :framework, :config, :driver, :thread
+    attr_accessor :state
+
+    def initialize(framework, config, driver)
+      self.state     = { }
+      self.framework = framework
+      self.config    = config
+      self.driver    = driver
+      self.thread    = framework.threads.spawn('OverwatchWorker', false) {
+        begin
+          self.start
+        rescue ::Exception => e
+          $stderr.puts "OverwatchWorker: #{e.class} #{e.to_s} #{e.backtrace}"
+        end
+
+        # Mark this worker as dead
+        self.thread = nil
+      }
+    end
+
+    def stop
+      return unless self.thread
+      self.thread.kill rescue nil
+      self.thread = nil
+    end
+
+    # Starts the overwatch worker and loops through the open sessions
+    # based on the defined timeframe.
+    # Primary error handling to the log file occurs here as well.
+    #
+    # @return [void] A useful return value is not expected here
+    def start
+      if self.config[:log]
+        self.driver.print_status("Overwatch is logging to #{self.config[:base]}. Use 'overwatch_start -h' for plugin information.")
+      else
+        self.driver.print_status("Overwatch started, logging is disabled. Use 'overwatch_start -h' for plugin information.")
+      end
+
+      bool_options = [ :autoroute, :reroute_stale, :kill_stale, :kill_stale_dup, :log ]
+      bool_options.each do |o|
+        self.config[o] = !!( self.config[o].to_s =~ /^[yt1]/i)
+      end
+
+      int_options = [ :freq, :route_timeout, :session_timeout, :session_dup_timeout ]
+      int_options.each do |o|
+        self.config[o] = self.config[o].to_i
+      end
+
+      # Setup the log file if logging is active.
+      ::FileUtils.mkdir_p(self.config[:base]) if self.config[:log]
+
+      loop do
+        framework.sessions.keys.each do |sid|
+
+          begin
+            if self.state[sid].nil? ||
+              (self.state[sid][:last_update] + self.config[:freq] < Time.now.to_f)
+              process(sid)
+            end
+
+          # Error handling - Yea, lots of stuff in here to keep errors from popping up in the console when sessions die or go stale. 
+          # We also skip sessions below that have more than three errors because they slow down processing.
+          rescue ::Exception => e
+            if e.class.to_s == "Rex::TimeoutError"  # Error when a session stops responding and the requests time out.
```

Comment:
I'll work on cleaning up the error handling per your suggestion.

A quick note: I'm still fairly new to Ruby and much of what I know in this language comes from looking at code that others have written. So any feedback or help is much appreciated.